### ISSUE TITLE
UI Buttons "Earlier" and "Later" in search view change search type

### ISSUE THREAD

**Issue by CryptidVulpes** [HSLdevcom/digitransit-ui#1354]
Searching for example "From: Kamppi" and "To: Otaniemi" (works with other searches as well) and changing results by the "Earlier" and "Later" buttons, your search type is also changed. This happens in beta.reittiopas.fi and also on www.reittiopas.fi

Steps to reproduce:
1. Search for Kamppi -> Otaniemi, Leaving, Today, 19:04 (https://beta.reittiopas.fi/reitti/Kamppi%2C%20Helsinki%3A%3A60.16902%2C24.931702/Otaniemi%2C%20Espoo%3A%3A60.187938%2C24.832845?arriveBy=false&time=1487178240)
2. Click "Earlier"
3. Observe "Leaving" has changed to "Arriving" and time has changed from "19:04" to "19:35"
4. Click "Later"
5. Observe "Arriving" has changed to "Leaving" and time has changed from "19:35" to "18:58"

Also the URL parameters change "Earlier"-button switches `arriveBy=false` -> `arriveBy=true` and vice-versa for "Later"-button.

My environment:
- OS: Windows 10
- Browser: Google Chrome version 56.0.2924.87

**Comment by tuukka** [comment]
Thanks for the report! I'm not sure how you'd expect/want it to work. Our reasoning behind the current logic has been that by defining "earlier" and "later" as "arriving before any of the current results" and "leaving after any of the current results" respectively, we always get a completely new set of results, and the skip size adjusts to the connections available.

The solution might be to hide this internal logic if it's too confusing: we would change the search type but not reflect it in the UI or URL parameters. Another direction to go might be to create a better time selector component if it's currently too clumsy to change the time by the amount which you have in mind (this currently depends on the browser that you use).

**Comment by CryptidVulpes** [comment]
Intuitively (for me at least), "earlier" means "earlier arrivals" or "earlier departures" depending on the original search, similar logic to "later". So I personally would expect something like this:

1. Leaving, Today, 19:04
2. Click "Earlier"
3. Leaving, Today, 18:54
4. Click "Later"
5. Leaving, Today, 19:04

OR

1. Arriving, Today, 19:04
2. Click "Earlier"
3. Arriving, Today, 18:54
4. Click "Later"
5. Arriving, Today, 19:04

... and so on, the increment is purely up to you of course :).

I think this is reflected in the UI as well, since your button ordering is like so: "Later", "Now", "Earlier". The same "design pattern" is visible in calendars and other timetable software with "<<", "Today", ">>" or similar.

**Comment by CryptidVulpes** [comment]
If you really, really, want to keep the current *behavior*, change to buttons to say "Earlier arrivals" and "Later departures" instead of just "Earlier" and "Later"

**Comment by hannesj** [comment]
Something that could be done instead of the logic would be to remember if the user came to the current state by clicking earlier or later and then go back to the initial state when going to the opposite direction in time, rather than doing a new search.

**Comment by nanonyme** [comment]
Btw, at least the old Reittiopas used to give *much* worse optimized search results if you queried by arrival instead of departure so touching those is a *really* bad thing unless something significantly changed in the routing algorithms.

**Comment by hannesj** [comment]
@nanonyme The search is based on a totally new routing engine based on https://github.com/opentripplanner/OpenTripPlanner rather than the old proprietary one. The searches work identically independent of the direction of the time in time in the algorithm.

Also the new engine searches for x best results for a single departure time, rather than searching for  the best result, adjusting the departure time to one minute after the departure time and searching for a single new result with that time and so on.

**Comment by Walther** [comment]
> Intuitively (for me at least), "earlier" means "earlier arrivals" or "earlier departures" depending on the original search, similar logic to "later".

I agree. Consider the two alternatives:
1. Pressing "earlier" moves the time window to an earlier point
2. Pressing "earlier" performs analysis of what the current search contained, adjusts search to match exact results that are earlier than any of the currently shown ones, **and changes your search type silently**.

Which one is more likely to be the expected behaviour by the user?

Additionally, even if just considering this on a more abstract level, having a button that does multiple things / affects multiple other UI choices seems rather complicated.

**Comment by Brownies** [comment]
> Intuitively (for me at least), "earlier" means "earlier arrivals" or "earlier departures" depending on the original search, similar logic to "later".  

+1. This is also how the old version of reittiopas works.

**Comment by tuukka** [comment]
> Which one is more likely to be the expected behaviour by the user?

Users report confusion if after pressing the button, they still see same trips.

> having a button that does multiple things / affects multiple other UI choices seems rather complicated.

For example, the Now button also has three functions: change search time to current time, change type to departure, update starting point to your current location if you haven't chosen another starting point.

> This is also how the old version of reittiopas works.

The search logic in the new one is intentionally different.

**Comment by jjonek** [comment]
I need to keep repeating that the current behavior is super confusing.

Searching for a route from Ruoholahti to Kamppi, leaving tomorrow 23:30 with only metro enabled the results suggests me to walk. Then after tapping *Earlier*, the first thing I notice is that the time was changed to 23.50. After that I realize that also the search mode was updated to *Arriving*. At this point I ignored the itinerary suggestion completely because of the confusion.

Now after reading these comments, I revisited that search. Indeed, the itinerary suggestions look sane. It's probably showing me the last metro connections of the day. But the completely illogical user experience really prevents me from trusting the results and forces me to go check the schedule from elsewhere.

Further, thinking about the user experience and the user trusting the results, I'm not sure if it's always the best idea to show a completely new set of results. Especially when the intervals are long. At least I would personally feel more confident if the last item after pressing *Earlier* would be the one that was the first earlier. Then I wouldn't need to speculate if the system missed some itineraries in between.
